File: data-structures/Queue.cpp
#include <iostream>

template <typename T> class Node
{
    public:
        T data;
        Node<T>* next;
};

template <typename T> class Queue
{
    private:
        friend class Node<T>;
        Node<T>* head = NULL;
        Node<T>* last = NULL;
        int n = 0; /* keeps track of the queue size */

    public:
        void enqueue(T data)
        {
            if (!isEmpty()) {
                Node<T>* newNode = new Node<T>();
                newNode->data = data;
                newNode->next = last->next;
                last->next = newNode;
                last = newNode;
            } else {
                last = head = new Node<T>();
                head->data = data;
                head->next = NULL;
            }
            n++;
        }

        T dequeue()
        {
            if (!isEmpty()) {
                Node<T>* oldhead = head;
                T data = oldhead->data;
                head = oldhead->next;
                delete oldhead;
                n--;
                return data;
            }

            return 0;
        }

        bool isEmpty()
        {
            return (head == NULL);
        }

        int size()
        {
            return n;
        }
        
        void print()
        {
            Node<T>* start;
            for (start = head; start != last; start = start->next)
                std::cout << start->data << " ";
            std::cout << last->data << std::endl;
        }

};

File: graphs/Queue.java


public class Queue {
    
    private class Node {
        public int val;
        public Node next = null;

        Node(int val) {
            this.val = val;
        }
    }

    private Node head = null;

    public void enqueue(int val) {
        var pivot = new Node(val);
        if (isEmpty()) {
            head = pivot;
        } else {
            pivot.next = head.next;
            head.next = pivot;
            head = head.next;
        }
    }

    public int dequeue() {
        var pivot = head.next;
        head.next = pivot.next;
        return pivot.val;
    }

    public boolean isEmpty() {
        return head == null;
    }

}
